### ISSUE TITLE
Running exporter as non-root? sudo for calls to smartctl?

### ISSUE THREAD

**Issue by frittentheke** [prometheus-community/smartctl_exporter#146]
I was wondering if it was really necessary to run the whole smartctl_exporter as root. This is even promoted via the provided systemd unit https://github.com/prometheus-community/smartctl_exporter/blob/master/systemd/smartctl_exporter.service

I know smartctl does seem to require this (https://github.com/smartmontools/smartmontools/issues/205). 
But maybe there is a way to not require for the exporter to run as root but to just call a privileged smartctl?
Is the good old sudo the right approach? Just by setting `smartctl.path` to `sudo smartctl.path` maybe?

**Comment by lahwaacz** [comment]
You can easily switch various [https://www.redhat.com/sysadmin/systemd-secure-services security features] of the systemd service to see what works and what not. I don't see an advantage in adding another layer with `sudo`...

**Comment by NiceGuyIT** [comment]
Adding `sudo` to the exporter is not going to happen.

As suggested by @lahwaacz, if you figure out how to configure systemd to improve the security, PRs are welcome!

**Comment by k0ste** [comment]
@frittentheke FYI [prometheus-smartctl-exporter.service in AUR](https://aur.archlinux.org/cgit/aur.git/tree/prometheus-smartctl-exporter.service?h=prometheus-smartctl-exporter)
If you can test it, tune it, PR it - it can be part of smartctl_exporter tarball

**Comment by frittentheke** [comment]
I see my provocation of creativity by just mentioning sudo worked :-). Thanks for taking the quest to reduce the privileges seriously and let's see if there is something that could work for most installs.

**Comment by DaAwesomeP** [comment]
The AUR service unit looks very promising! Unfortunately I am not able to test/pull at this time.

The full documentation of these options can he found here: https://www.freedesktop.org/software/systemd/man/systemd.exec.html

Note that not all distros ship the latest Systemd, so newer options may result in warnings in the journal. However, in my experience it is backward-compatible without issue (you may not get _all_ of the security features but still as many as possible).

**Comment by subnix** [comment]
I'm looking for a secure way to run the exporter too. These privileges are required for smartctl:
1. `CAP_SYS_RAWIO` capability to be able to access raw devices;
2. `disk` user group, which owns all the disks.

So you can add these lines to the systemd unit:
```
[Service]
...
CapabilityBoundingSet=CAP_SYS_RAWIO
AmbientCapabilities=CAP_SYS_RAWIO
Group=disk
...
```

But this set of privileges may lead to privilege escalation since root filesystem is accessible for writing. So `sudo` with limiting sudoers configuration still looks more secure.

**Comment by zar3bski** [comment]
Agreed with @subnix : Once you introduce `CAP_SYS_RAWIO`, none of the hardening (e.g. `DeviceAllow`) methods could restrict the ability for the service to perform arbitrary write on any of the block device. In this end, you end up with an HTTP service listening to the external world with root abilities on the block devices so I find that the sudo/doas whitelist approach adds something IMHO

**Comment by k0ste** [comment]
> The AUR service unit looks very promising!

[prometheus-smartctl-exporter.service](https://gitlab.archlinux.org/archlinux/packaging/packages/prometheus-smartctl-exporter/-/blob/main/prometheus-smartctl-exporter.service?ref_type=heads) in Archlinux upstream 🎉
